README.md
# NAME

Plack::Response::AutoEncode - Plack response with automatic encoding feature

# SYNOPSIS

in your PSGI application

    use utf8;
    use Plack::Response::AutoEncode;

    my $app = sub {
        my $body = '私は日本人です。';
        my $res  = Plack::Response::AutoEncode->new(200, ['text/html; charset=UTF-8'], [$body]);
        $res->finalize;
    };

# DESCRIPTION

__DO NOT USE THIS MODULE!!!!__ This module broke interface of Plack::Response.

Plack::Response::AutoEncode is subclass of Plack::Response.

When application returns a response that contains "text/\*" in Content-Type header, encode automatically each unencoded content by charset that is in Content-Type header.

You __MUST__ set content-body data in perl-string.

For example. If you want to response with Shift\_JIS encoding, you can it as followings.

    use utf8;
    use Plack::Response::AutoEncode;

    my $app = sub {
        my $body = '私は日本人です。';
        ### like as s|UTF-8|Shift_JIS|;
        my $res  = Plack::Response::AutoEncode->new(200, ['text/html; charset=Shift_JIS'], [$body]);
        $res->finalize;
    };

# LICENSE

Copyright (C) ytnobody.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

ytnobody <ytnobody@gmail.com>

moznion


Changes
Revision history for Perl extension Plack-Response-AutoEncode

{{$NEXT}}

0.02 2013-11-15T08:30:50Z
    - require perl-string in content-body
    - force valid charset

0.01 2013-11-15T06:49:18Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Plack-Response-AutoEncode',
    module_name     => 'Plack::Response::AutoEncode',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Plack/Response/AutoEncode.pm
package Plack::Response::AutoEncode;
use strict;
use warnings;
use parent 'Plack::Response';
use Carp ();
use Email::MIME::ContentType () ;
use Encode;
our $VERSION = "0.02";

sub finalize {
    my $self = shift;
    if ($self->content_type =~ /^text\//) {
        my $data    = Email::MIME::ContentType::parse_content_type($self->header('Content-Type'));
        my $charset = $data->{attributes}{charset};
        if ($charset) {
            my $content = ref($self->content) eq 'ARRAY' ?
                [ map { $self->_encode_gracefully($charset, $_) } @{$self->content} ] :
                $self->_encode_gracefully( $self->content )
            ;
            $self->content($content);

            if ($self->{encoding}) {
                $data->{attributes}{charset} = $self->{encoding}->mime_name;
            }

            $self->headers->{'content-type'} = $self->_build_content_type($data);
        }
    }
    $self->SUPER::finalize(@_);
}

sub _encode_gracefully {
    my ($self, $charset, $str) = @_;

    $self->{encoding} = Encode::find_encoding($charset);
    unless ($self->{encoding}) {
        Carp::carp qq![Error] Invalid charset was detected ("$charset")!;

        # $str is must perl-string. Return it that is encoded by UTF-8 for the time being unless defined 'encoding'...
        return Encode::encode('utf8', $str);
    }

    $self->{encoding}->encode($str);
}

sub _build_content_type {
    my ($self, $data) = @_;

    my $content_type_str = "$data->{discrete}/$data->{composite}";
    for my $k (keys %{$data->{attributes}}) {
        $content_type_str .= ";$k=$data->{attributes}->{$k}";
    }
    return $content_type_str;
}
1;
__END__

=encoding utf-8

=head1 NAME

Plack::Response::AutoEncode - Plack response with automatic encoding feature

=head1 SYNOPSIS

in your PSGI application

    use utf8;
    use Plack::Response::AutoEncode;

    my $app = sub {
        my $body = '私は日本人です。';
        my $res  = Plack::Response::AutoEncode->new(200, ['text/html; charset=UTF-8'], [$body]);
        $res->finalize;
    };

=head1 DESCRIPTION

B<DO NOT USE THIS MODULE!!!!> This module broke interface of Plack::Response.

Plack::Response::AutoEncode is subclass of Plack::Response.

When application returns a response that contains "text/*" in Content-Type header, encode automatically each unencoded content by charset that is in Content-Type header.

You B<MUST> set content-body data in perl-string.

For example. If you want to response with Shift_JIS encoding, you can it as followings.

    use utf8;
    use Plack::Response::AutoEncode;

    my $app = sub {
        my $body = '私は日本人です。';
        ### like as s|UTF-8|Shift_JIS|;
        my $res  = Plack::Response::AutoEncode->new(200, ['text/html; charset=Shift_JIS'], [$body]);
        $res->finalize;
    };

=head1 LICENSE

Copyright (C) ytnobody.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

ytnobody E<lt>ytnobody@gmail.comE<gt>

moznion

=cut

